jobsdb/unionQuery_test.go

```diff
@@ -111,3 +113,230 @@ func TestMultiTenantHandleT_GetAllJobs(t *testing.T) {
+BenchmarkUnionQuery/ParallelQuery-10000-8                       	                                 11816298834 ns/op
+
+BenchmarkUnionQuery/UnionQuery-100000-8                         	                                 143067570417 ns/op
+BenchmarkUnionQuery/Concurrent_Union_Query__-100000_-16-8       	                 200310106917 ns/op
+BenchmarkUnionQuery/Concurrent_Union_Query__-100000_-32-8       	                 409845463834 ns/op
+BenchmarkUnionQuery/ParallelQuery-100000-8                      	                                  118793139708 ns/op
+*/
+
+// BenchmarkUnionQuery is a benchmark for UnionQuery with 20 different workspaces
+// and jobSize is the total number of jobs split across the workspaces
+func BenchmarkUnionQuery(b *testing.B) {
+	_ = startPostgres(b)
+	jobSize := []int{10000}
+	//concurrencies := []int{16, 32}
+	partitionTypes := []string{"HASH"}
+	tablePrefixes := []string{"rt"}
+	workspacesCounts := []int{30, 50, 100, 150, 200}
+	count := 0
+	for _, size := range jobSize {
+		for _, workspacesCount := range workspacesCounts {
+			workspaces := generateWorkspaces(workspacesCount)
+			workspaceSize := (size + workspacesCount - 1) / workspacesCount
+			workspaceDistribution := []string{}
+			var j int
+			for i, workspace := range workspaces {
+				for j = 0; j < workspaceSize && i*j < size; j++ {
+					workspaceDistribution = append(workspaceDistribution, workspace)
+				}
+				if i*j >= size {
+					break
+				}
+			}
+			jobs, testWorkspacePickup := generateJobs(size, count, workspaceDistribution)
+			count += size
+			for _, partitionType := range partitionTypes {
+				b.Setenv("RSERVER_JOBS_DB_ENABLE_WRITER_QUEUE", "true")
+				b.Setenv("RSERVER_JOBS_DB_ENABLE_READER_QUEUE", "true")
+				for _, tablePrefix := range tablePrefixes {
+					//jobsDb1 := MultiTenantHandleT{HandleT: &HandleT{}}
+					//err := jobsDb1.Setup(ReadWrite, true, tablePrefix, true, []prebackup.Handler{}, fileuploader.NewDefaultProvider())
+					//require.NoError(b, err)
+					//err = jobsDb1.Store(context.Background(), jobs)
+					//require.NoError(b, err)
+					//
+					//b.Run(fmt.Sprintf("UnionQuery-%d-%s", size, tablePrefix), func(b *testing.B) {
+					//	benchmarkUnionQuery(b, jobsDb1, testWorkspacePickup, "testCustomVal")
+					//})
+					//
+					////for _, concurrency := range concurrencies {
+					////	b.Run(fmt.Sprintf("ConcurrentUnionQuery-%d-%d-%s", size, concurrency, tablePrefix), func(b *testing.B) {
+					////		benchmarkConcurrentUnionQuery(b, jobsDb1, testWorkspacePickup, "testCustomVal", concurrency)
+					////	})
+					////}
+					//
+					//jobsDb1.TearDown()
+
+					maxDSSize := 5
+					jobsDb2 := MultiTenantLegacy{HandleT: &HandleT{
+						MaxDSSize: &maxDSSize,
+					}}
+					err := jobsDb2.Setup(ReadWrite, true, tablePrefix, true, []prebackup.Handler{}, fileuploader.NewDefaultProvider(), partitionType)
+					require.NoError(b, err)
+					err = jobsDb2.Store(context.Background(), jobs)
+					require.NoError(b, err)
```

Comment:
@cisse21 @BonapartePC beware that we're still writing first and then reading. in production we're doing both at the same time so the benchmark won't highlight the difference in performance that we would get in a real scenario.

Reply:
Let's have a minimum number of records that are already in the table when the benchmark starts and then a write frequency (e.g. start with 100k rows, then start benchmarking the reads but have X writes per second while the reads are running). We'll have the same number of writes per second when benchmarking different types of queries so the benchmark will still be valid.